mm2src/coins/lightning/ln_filesystem_persister.rs

```diff
@@ -96,43 +72,118 @@ impl LightningFilesystemPersister {
-{
-    fn persist_manager(&self, channel_manager: &ChannelManager) -> Result<(), std::io::Error> {
-        FilesystemPersister::persist_manager(self.0.main_path().display().to_string(), channel_manager)?;
-        if let Some(backup_path) = self.0.backup_path() {
-            let file = fs::OpenOptions::new()
-                .create(true)
-                .write(true)
-                .truncate(true)
-                .open(backup_path)?;
-            channel_manager.write(&mut BufWriter::new(file))?;
+    /// Read `ChannelMonitor`s from disk.
+    pub fn read_channelmonitors<Signer: Sign, K: Deref>(
+        &self,
+        keys_manager: K,
+    ) -> Result<Vec<(BlockHash, ChannelMonitor<Signer>)>, std::io::Error>
+    where
+        K::Target: KeysInterface<Signer = Signer> + Sized,
+    {
+        let path = self.monitors_path();
+        if !path.exists() {
+            return Ok(Vec::new());
         }
-        Ok(())
+        let mut res = Vec::new();
+        for file_option in fs::read_dir(path).unwrap() {
+            let file = file_option.unwrap();
+            let owned_file_name = file.file_name();
+            let filename = owned_file_name.to_str();
+            if filename.is_some() && filename.unwrap() == "checkval" {
+                continue;
+            }
+            if filename.is_none() || !filename.unwrap().is_ascii() || filename.unwrap().len() < 65 {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::InvalidData,
+                    "Invalid ChannelMonitor file name",
+                ));
+            }
+            if filename.unwrap().ends_with(".tmp") {
+                // If we were in the middle of committing an new update and crashed, it should be
+                // safe to ignore the update - we should never have returned to the caller and
+                // irrevocably committed to the new state in any way.
+                continue;
+            }
+
+            let txid = Txid::from_hex(filename.unwrap().split_at(64).0);
+            if txid.is_err() {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::InvalidData,
+                    "Invalid tx ID in filename",
+                ));
+            }
+
+            let index = filename.unwrap().split_at(65).1.parse::<u16>();
+            if index.is_err() {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::InvalidData,
+                    "Invalid tx index in filename",
+                ));
+            }
+
+            let contents = fs::read(&file.path())?;
+            let mut buffer = Cursor::new(&contents);
+            match <(BlockHash, ChannelMonitor<Signer>)>::read(&mut buffer, &*keys_manager) {
+                Ok((blockhash, channel_monitor)) => {
+                    if channel_monitor.get_funding_txo().0.txid != txid.unwrap()
+                        || channel_monitor.get_funding_txo().0.index != index.unwrap()
```

Comment:
This is also not safe

Reply:
There is a check for if `txid` and `index` are errors or not above https://github.com/KomodoPlatform/atomicDEX-API/blob/a7e95e97891531b2c4c9331f837232d83865fd6a/mm2src/coins/lightning/ln_filesystem_persister.rs#L134-L148 But refecatoring this to return the value when they are not an error better.

Reply:
Done